File: context-manager/src/core/lease.rs
//! Compaction mutual exclusion (context-manager.md § State): a
//! `{nonce, ts}` claim under scope `context_lease`, keyed by
//! `options.lease_key` (e.g. a session id) or a hash of the message
//! set. TTL is enforced by readers — a claim older than the TTL reads
//! as free, folding crash recovery into acquisition. Ported from
//! harness `runtime/lease.ts`.

use sha2::{Digest, Sha256};

use crate::ports::{Clock, LeaseRecord, LeaseStore};
use crate::types::AgentMessage;

/// The single lease scope this worker writes (a subdirectory under the
/// configured `lease_dir`).
pub const LEASE_SCOPE: &str = "context_lease";

/// Default lease key: sha256 over the serialized message set, so two
/// callers compacting the same logical history contend even without an
/// explicit `lease_key`, while different histories never block each
/// other. sha2 (not a std hasher) keeps the key stable across
/// processes and Rust versions.
pub fn default_lease_key(messages: &[AgentMessage]) -> String {
    let mut hasher = Sha256::new();
    for message in messages {
        hasher.update(serde_json::to_string(message).unwrap_or_default());
        hasher.update([0u8]); // message separator
    }
    format!("{:x}", hasher.finalize())
}

fn is_active(record: Option<&LeaseRecord>, now_ms: i64, ttl_ms: i64) -> bool {
    record.is_some_and(|r| now_ms - r.ts < ttl_ms)
}

/// Try to acquire the lease. Returns the nonce to release with, or
/// `None` when another holder owns a live claim. Store failures read
/// as "not acquired" (never a win) — an outage must not let every
/// contender through, and a transient `busy` is retryable.
pub async fn acquire(
    store: &dyn LeaseStore,
    clock: &dyn Clock,
    key: &str,
    ttl_ms: i64,
) -> Option<String> {
    let now = clock.now_ms();
    match store.get(LEASE_SCOPE, key).await {
        Ok(existing) if is_active(existing.as_ref(), now, ttl_ms) => return None,
        Ok(_) => {}
        Err(e) => {
            tracing::warn!(error = %e, key, "lease get failed; treating as busy");
            return None;
        }
    }

    let nonce = format!("{}-{}", std::process::id(), uuid::Uuid::new_v4().simple());
    let claim = LeaseRecord {
        nonce: nonce.clone(),
        ts: now,
    };
    let prior = match store.swap(LEASE_SCOPE, key, claim).await {
        Ok(prior) => prior,
        Err(e) => {
            tracing::warn!(error = %e, key, "lease swap failed; never treating as a win");
            return None;
        }
    };

    // We clobbered a still-valid claim (always someone else's — our
    // nonce is fresh): restore it and bow out.
    if is_active(prior.as_ref(), now, ttl_ms) {
        if let Err(e) = store.set(LEASE_SCOPE, key, prior).await {
            tracing::warn!(error = %e, key, "lease restore failed");
        }
        return None;
    }
    Some(nonce)
}

/// Release only our own claim: a holder that lost its lease to TTL
/// takeover must not clear the new holder's claim.
pub async fn release(store: &dyn LeaseStore, key: &str, nonce: &str) {
    match store.get(LEASE_SCOPE, key).await {
        Ok(Some(stored)) if stored.nonce == nonce => {
            if let Err(e) = store.set(LEASE_SCOPE, key, None).await {
                tracing::warn!(error = %e, key, "lease clear failed");
            }
        }
        Ok(_) => {}
        Err(e) => tracing::warn!(error = %e, key, "lease release read failed"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn default_key_is_stable_and_input_sensitive() {
        let a: AgentMessage = serde_json::from_value(json!({
            "role": "user", "content": [{ "type": "text", "text": "hi" }], "timestamp": 1
        }))
        .unwrap();
        let b: AgentMessage = serde_json::from_value(json!({
            "role": "user", "content": [{ "type": "text", "text": "bye" }], "timestamp": 1
        }))
        .unwrap();

        let key_a1 = default_lease_key(std::slice::from_ref(&a));
        let key_a2 = default_lease_key(std::slice::from_ref(&a));
        let key_b = default_lease_key(std::slice::from_ref(&b));
        assert_eq!(key_a1, key_a2, "same messages must contend");
        assert_ne!(
            key_a1, key_b,
            "different messages must not block each other"
        );
        assert_eq!(key_a1.len(), 64, "sha256 hex");
    }

    #[test]
    fn concatenation_is_not_ambiguous() {
        // ["ab"] vs ["a", "b"]-style boundary games must not collide;
        // the separator byte keeps message boundaries in the hash.
        let one: Vec<AgentMessage> = serde_json::from_value(json!([
            { "role": "user", "content": [{ "type": "text", "text": "ab" }], "timestamp": 1 }
        ]))
        .unwrap();
        let two: Vec<AgentMessage> = serde_json::from_value(json!([
            { "role": "user", "content": [{ "type": "text", "text": "a" }], "timestamp": 1 },
            { "role": "user", "content": [{ "type": "text", "text": "b" }], "timestamp": 1 }
        ]))
        .unwrap();
        assert_ne!(default_lease_key(&one), default_lease_key(&two));
    }

    #[test]
    fn activity_is_ttl_relative() {
        let rec = LeaseRecord {
            nonce: "n".into(),
            ts: 1_000,
        };
        assert!(is_active(Some(&rec), 1_500, 1_000)); // 500ms old, 1s TTL
        assert!(!is_active(Some(&rec), 2_500, 1_000)); // expired
        assert!(!is_active(None, 0, 1_000));
    }
}

File: harness/src/runtime/lease.ts
/**
 * Single-writer lease over `state::update` — the only atomic read-modify-write
 * iii exposes (there's no native lock). Acquire writes a `{nonce, ts}` claim and
 * reads the prior value in one atomic op, so exactly one concurrent acquirer
 * sees a free/expired prior and wins; a claim older than `ttlMs` reads inactive,
 * folding crash recovery into the same op. session-lease is a thin (scope, ttl)
 * adapter over this.
 */

import { randomUUID } from 'node:crypto';
import type { ISdk } from './iii.js';
import { stateGet, stateSet, stateUpdate } from './state.js';

export type LeaseRecord = { nonce: string; ts: number };

let counter = 0;

export function mintLeaseNonce(): string {
  return `${process.pid}-${process.hrtime.bigint()}-${counter++}-${randomUUID().slice(0, 8)}`;
}

function isLeaseActive(rec: LeaseRecord | null | undefined, nowMs: number, ttlMs: number): boolean {
  return rec != null && nowMs - rec.ts < ttlMs;
}

/** Returns the nonce to release with, or null if another holder owns a valid lease. */
export async function acquireLease(
  iii: ISdk,
  scope: string,
  key: string,
  ttlMs: number,
): Promise<string | null> {
  const now = Date.now();
  if (isLeaseActive(await stateGet<LeaseRecord>(iii, scope, key), now, ttlMs)) return null;

  const nonce = mintLeaseNonce();
  const result = await stateUpdate<LeaseRecord>(iii, scope, key, [
    { type: 'set', path: '', value: { nonce, ts: now } },
  ]);
  // A null envelope means the atomic write failed — never treat that as a win,
  // or an outage lets every contender through.
  if (!result) return null;

  // We clobbered a still-valid claim (always someone else's — our nonce is
  // fresh): restore it and bow out.
  if (isLeaseActive(result.old_value, now, ttlMs)) {
    await stateSet(iii, scope, key, result.old_value);
    return null;
  }
  return nonce;
}

export async function releaseLease(
  iii: ISdk,
  scope: string,
  key: string,
  nonce: string,
): Promise<void> {
  const stored = await stateGet<LeaseRecord>(iii, scope, key);
  if (stored?.nonce === nonce) await stateSet(iii, scope, key, null);
}

/** Poll {@link acquireLease} with exponential backoff until acquired or timed out. */
export async function acquireLeaseWithWait(
  iii: ISdk,
  scope: string,
  key: string,
  ttlMs: number,
  totalTimeoutMs: number,
): Promise<string | null> {
  const deadline = Date.now() + totalTimeoutMs;
  let backoff = 50;
  while (true) {
    const nonce = await acquireLease(iii, scope, key, ttlMs);
    if (nonce) return nonce;
    if (Date.now() + backoff > deadline) return null;
    await new Promise((r) => setTimeout(r, backoff));
    backoff = Math.min(backoff * 2, 500);
  }
}
